File: sdk/typescript/src/api/saved-playlists.ts
import type { HttpTransport } from '../transport.js';
import type { SavedPlaylist, SavedPlaylistFolder } from '../types.js';

export interface CreatePlaylistInput {
  name: string;
  description?: string;
  image?: string;
  folderId?: string;
  trackIds?: string[];
}

export interface UpdatePlaylistInput {
  name: string;
  description?: string;
  image?: string;
  folderId?: string;
}

export class SavedPlaylistsApi {
  constructor(private readonly http: HttpTransport) {}

  async list(folderId?: string): Promise<SavedPlaylist[]> {
    const data = await this.http.execute<{ savedPlaylists: SavedPlaylist[] }>(/* GraphQL */ `
      query SavedPlaylists($folderId: String) {
        savedPlaylists(folderId: $folderId) {
          id name description image folderId trackCount createdAt updatedAt
        }
      }
    `, { folderId });
    return data.savedPlaylists;
  }

  async get(id: string): Promise<SavedPlaylist | null> {
    const data = await this.http.execute<{ savedPlaylist: SavedPlaylist | null }>(/* GraphQL */ `
      query SavedPlaylist($id: String!) {
        savedPlaylist(id: $id) {
          id name description image folderId trackCount createdAt updatedAt
        }
      }
    `, { id });
    return data.savedPlaylist;
  }

  async trackIds(playlistId: string): Promise<string[]> {
    const data = await this.http.execute<{ savedPlaylistTrackIds: string[] }>(/* GraphQL */ `
      query SavedPlaylistTrackIds($playlistId: String!) {
        savedPlaylistTrackIds(playlistId: $playlistId)
      }
    `, { playlistId });
    return data.savedPlaylistTrackIds;
  }

  async create(input: CreatePlaylistInput): Promise<SavedPlaylist> {
    const data = await this.http.execute<{ createSavedPlaylist: SavedPlaylist }>(/* GraphQL */ `
      mutation CreateSavedPlaylist(
        $name: String!, $description: String, $image: String,
        $folderId: String, $trackIds: [String!]
      ) {
        createSavedPlaylist(
          name: $name, description: $description, image: $image,
          folderId: $folderId, trackIds: $trackIds
        ) {
          id name description image folderId trackCount createdAt updatedAt
        }
      }
    `, input);
    return data.createSavedPlaylist;
  }

  async update(id: string, input: UpdatePlaylistInput): Promise<void> {
    await this.http.execute(/* GraphQL */ `
      mutation UpdateSavedPlaylist(
        $id: String!, $name: String!, $description: String, $image: String, $folderId: String
      ) {
        updateSavedPlaylist(
          id: $id, name: $name, description: $description, image: $image, folderId: $folderId
        )
      }
    `, { id, ...input });
  }

  async delete(id: string): Promise<void> {
    await this.http.execute(/* GraphQL */ `
      mutation DeleteSavedPlaylist($id: String!) { deleteSavedPlaylist(id: $id) }
    `, { id });
  }

  async addTracks(playlistId: string, trackIds: string[]): Promise<void> {
    await this.http.execute(/* GraphQL */ `
      mutation AddTracksToSavedPlaylist($playlistId: String!, $trackIds: [String!]!) {
        addTracksToSavedPlaylist(playlistId: $playlistId, trackIds: $trackIds)
      }
    `, { playlistId, trackIds });
  }

  async removeTrack(playlistId: string, trackId: string): Promise<void> {
    await this.http.execute(/* GraphQL */ `
      mutation RemoveTrackFromSavedPlaylist($playlistId: String!, $trackId: String!) {
        removeTrackFromSavedPlaylist(playlistId: $playlistId, trackId: $trackId)
      }
    `, { playlistId, trackId });
  }

  async play(playlistId: string): Promise<void> {
    await this.http.execute(/* GraphQL */ `
      mutation PlaySavedPlaylist($playlistId: String!) { playSavedPlaylist(playlistId: $playlistId) }
    `, { playlistId });
  }

  // ---------------------------------------------------------------------------
  // Folders
  // ---------------------------------------------------------------------------

  async folders(): Promise<SavedPlaylistFolder[]> {
    const data = await this.http.execute<{ playlistFolders: SavedPlaylistFolder[] }>(/* GraphQL */ `
      query PlaylistFolders {
        playlistFolders { id name createdAt updatedAt }
      }
    `);
    return data.playlistFolders;
  }

  async createFolder(name: string): Promise<SavedPlaylistFolder> {
    const data = await this.http.execute<{ createPlaylistFolder: SavedPlaylistFolder }>(/* GraphQL */ `
      mutation CreatePlaylistFolder($name: String!) {
        createPlaylistFolder(name: $name) { id name createdAt updatedAt }
      }
    `, { name });
    return data.createPlaylistFolder;
  }

  async deleteFolder(id: string): Promise<void> {
    await this.http.execute(/* GraphQL */ `
      mutation DeletePlaylistFolder($id: String!) { deletePlaylistFolder(id: $id) }
    `, { id });
  }
}

File: sdk/python/src/rockbox_sdk/api/saved_playlists.py
"""Persistent ("saved") playlists and their folders."""

from __future__ import annotations

from dataclasses import asdict, dataclass

from ..transport import HttpTransport
from ..types import SavedPlaylist, SavedPlaylistFolder

_PLAYLIST_FIELDS = "id name description image folderId trackCount createdAt updatedAt"
_FOLDER_FIELDS = "id name createdAt updatedAt"


@dataclass
class CreatePlaylistInput:
    name: str
    description: str | None = None
    image: str | None = None
    folder_id: str | None = None
    track_ids: list[str] | None = None


@dataclass
class UpdatePlaylistInput:
    name: str
    description: str | None = None
    image: str | None = None
    folder_id: str | None = None


def _camelize(d: dict[str, object]) -> dict[str, object]:
    """folder_id → folderId, track_ids → trackIds, etc."""
    out: dict[str, object] = {}
    for k, v in d.items():
        if v is None:
            continue
        parts = k.split("_")
        camel = parts[0] + "".join(p.title() for p in parts[1:])
        out[camel] = v
    return out


class SavedPlaylistsApi:
    def __init__(self, http: HttpTransport) -> None:
        self._http = http

    async def list(self, folder_id: str | None = None) -> list[SavedPlaylist]:
        data = await self._http.execute(
            "query SavedPlaylists($folderId: String) "
            f"{{ savedPlaylists(folderId: $folderId) {{ {_PLAYLIST_FIELDS} }} }}",
            {"folderId": folder_id},
        )
        return [SavedPlaylist.model_validate(p) for p in data.get("savedPlaylists", [])]

    async def get(self, id: str) -> SavedPlaylist | None:
        data = await self._http.execute(
            "query SavedPlaylist($id: String!) "
            f"{{ savedPlaylist(id: $id) {{ {_PLAYLIST_FIELDS} }} }}",
            {"id": id},
        )
        raw = data.get("savedPlaylist")
        return SavedPlaylist.model_validate(raw) if raw is not None else None

    async def track_ids(self, playlist_id: str) -> list[str]:
        data = await self._http.execute(
            "query SavedPlaylistTrackIds($playlistId: String!) "
            "{ savedPlaylistTrackIds(playlistId: $playlistId) }",
            {"playlistId": playlist_id},
        )
        return list(data.get("savedPlaylistTrackIds", []))

    async def create(self, input: CreatePlaylistInput) -> SavedPlaylist:
        data = await self._http.execute(
            "mutation CreateSavedPlaylist("
            "$name: String!, $description: String, $image: String, "
            "$folderId: String, $trackIds: [String!]) "
            "{ createSavedPlaylist("
            "name: $name, description: $description, image: $image, "
            "folderId: $folderId, trackIds: $trackIds) "
            f"{{ {_PLAYLIST_FIELDS} }} }}",
            _camelize(asdict(input)),
        )
        return SavedPlaylist.model_validate(data["createSavedPlaylist"])

    async def update(self, id: str, input: UpdatePlaylistInput) -> None:
        await self._http.execute(
            "mutation UpdateSavedPlaylist("
            "$id: String!, $name: String!, $description: String, "
            "$image: String, $folderId: String) "
            "{ updateSavedPlaylist("
            "id: $id, name: $name, description: $description, "
            "image: $image, folderId: $folderId) }",
            {"id": id, **_camelize(asdict(input))},
        )

    async def delete(self, id: str) -> None:
        await self._http.execute(
            "mutation DeleteSavedPlaylist($id: String!) { deleteSavedPlaylist(id: $id) }",
            {"id": id},
        )

    async def add_tracks(self, playlist_id: str, track_ids: list[str]) -> None:
        await self._http.execute(
            "mutation AddTracksToSavedPlaylist($playlistId: String!, $trackIds: [String!]!) "
            "{ addTracksToSavedPlaylist(playlistId: $playlistId, trackIds: $trackIds) }",
            {"playlistId": playlist_id, "trackIds": track_ids},
        )

    async def remove_track(self, playlist_id: str, track_id: str) -> None:
        await self._http.execute(
            "mutation RemoveTrackFromSavedPlaylist($playlistId: String!, $trackId: String!) "
            "{ removeTrackFromSavedPlaylist(playlistId: $playlistId, trackId: $trackId) }",
            {"playlistId": playlist_id, "trackId": track_id},
        )

    async def play(self, playlist_id: str) -> None:
        await self._http.execute(
            "mutation PlaySavedPlaylist($playlistId: String!) "
            "{ playSavedPlaylist(playlistId: $playlistId) }",
            {"playlistId": playlist_id},
        )

    # --- folders --------------------------------------------------------

    async def folders(self) -> list[SavedPlaylistFolder]:
        data = await self._http.execute(
            f"query PlaylistFolders {{ playlistFolders {{ {_FOLDER_FIELDS} }} }}"
        )
        return [SavedPlaylistFolder.model_validate(f) for f in data.get("playlistFolders", [])]

    async def create_folder(self, name: str) -> SavedPlaylistFolder:
        data = await self._http.execute(
            "mutation CreatePlaylistFolder($name: String!) "
            f"{{ createPlaylistFolder(name: $name) {{ {_FOLDER_FIELDS} }} }}",
            {"name": name},
        )
        return SavedPlaylistFolder.model_validate(data["createPlaylistFolder"])

    async def delete_folder(self, id: str) -> None:
        await self._http.execute(
            "mutation DeletePlaylistFolder($id: String!) { deletePlaylistFolder(id: $id) }",
            {"id": id},
        )
